ENH: added association measurements Cramer's V, Tschuprow's T and Phi to scipy.contingency

dtemkin opened the pull request:
Added measures of association for nominal data to contingency stats module. Measures added include Tschuprow's T, Cramer's V and Phi. Measures are individual specified by values of the 'stat' argument (as 'T', 'V' or 'phi', respectively) within function scipy.stats.contingency.association_test.

These measures can provide useful information about variable independence to those performing chi-squared tests or categorical data analysis in general. Each of the measures is empirically estimated using the chi-squared statistic which, is automatically determined if the 'chi2_stat' argument is not specified. Also included is an option to apply a proposed method of bias correction developed by Bergsma (2013) [full reference included in code docs]. Lastly due to the measures requiring the specification of the number of rows and columns in the contingency table, a private function was added to reshape the any array that exceeded two dimensions to an N x N array. 

Tests were created for all included measures both with and without bias correction.

person142 commented:
Note: the Travis CI failures from `ndimage` are due to https://github.com/scipy/scipy/issues/8145. You don't need to try and fix them here. It's fine if the CI failures aren't due to *your* code.

dtemkin commented:
@person142 Thanks for letting me know.

person142 commented:
> Will the review process for my code be put on hold until there is a fix for this issue?

No, we still merge code in situations like this when it's clear the failures are unrelated.

dtemkin commented:
@person142 Now that my PR is passing all the checks, I am just wondering if there is anything I can do to help the review process along?

person142 commented:
I can't really think of anything. I am not knowledgeable enough on the topic to review it myself, so you'll have to wait until a reviewer with more stats knowledge has some free time. If a while goes by with no word from anyone (a couple of weeks or so) then you can bump. (Sometimes PRs have a way of slipping from the mind...)

dtemkin commented:
@josef-pkt 
1) The array reshape is necessary to handle an ndarray because both functions explicitly require number of rows and number of columns. Since the number of rows/columns is not so easily determined in an array that is, for example, 4x4x2x2, it was necessary to restructure the array to fit an NxN format. In reality, I can't imagine a circumstance in which a contingency table would have more than two dimensions but I was just looking ahead for users with strange datasets. 

2) What about just 'association' for the name of the function? I agree that '_test' isn't exactly right - I went back and forth myself.

3) I can't find a name for them either although, they may have some practical applications outside of contingency tables. However, given their integral nature to the functionality of contingency tables I guess this is the place for them. That is why I put the measures of association here. Because the formulae use the chi-squared statistic. If this is incorrect let me know.

I updated the script with the changes you recommended. Thanks for looking it over.

josef-pkt commented:
to 1: If there is no clear interpretation and definition for larger than 2-dimensional tables, then raising a ValueError might be more appropriate.
E.g. if I have a stratified 4x4xk contingency table, then I might want to have a separate or average association for each strata, ie. vectorized/looped over the extra dimensions. (I'm just guessing.)
If we have independence in a arbitrary k-dimensional contingency table as reference case, then it might be an appropriate overall association measure. (Also just guessing.)

to 3. Having `association` in the stats name space is fine and the proper location. However, IIUC, the intention was not to have margins and expected_freq in an unqualified stats namespace.
@WarrenWeckesser 

`association` might be a bit too general given that all correlation measures are measures of association, and this measures are only defined for nominal data. (I don't have a strong opinion about this.)

dtemkin commented:
@josef-pkt Sorry for the belated response. Your comment about stratified analyses made a lot of sense to me so I changed a lot in the function. I changed the function so that it looks for fixed size 2d arrays inside a Nd array. Then does a test for association on all the subarrays - returning an array of association stats as the output. The returned array is in the form of (N-1)d . I also changed the namespace so that the "association" function is bound to contingency. Let me know what you think.

dtemkin commented:
@josef-pkt P.S. The final check from Travis failed due to a system configuration error.

rosstripi commented:
Have these changes been merge into master yet? It's been two years, and post-hoc tests like Cramer's would be very helpful.

rgommers commented:
No, unfortunately not, it seems to have fallen off the radar. Comments from the first round of review seem addressed, and the code looks like it's in decent shape. So I've added a milestone 1.4.0 (next release), so that we will look at it again before then.

`stats.contingency` was written by @WarrenWeckesser IIRC, so hopefully he has some time to look at this.

dtemkin commented:
@rosstripi @rgommers Glad to see there is renewed interest in my forgotten PR . Since my fork was very old I thought it made sense to merge before continuing and fix any errors that arise. I also added Pearsons Contingency Coefficient for solidarity. If you have any questions, comments or suggestions please don't hesitate.

WarrenWeckesser commented:
I'll take a look, but I might not get to it until this weekend.

dtemkin commented:
@WarrenWeckesser It would be wonderful if you are able to find time to take a look. I have no idea how to fix the odd travis-ci errors - even though they did seem to occur after I added an exception handle verifying the data input type, but I cant imagine how that causes the errors being reported.  Lastly, I am pretty sure the circle-ci is not an issue with my commit but any guidance would be much appreciated.

WarrenWeckesser commented:
@dtemkin, looks like I'll need a few more days to get up to speed on these tests, and to understand your proposed API.   After a quick look, I suspect I'll have similar questions to those asked by @josef-pkt.  It looks like all these tests are for two-way (i.e. "R x C" or two-dimensional) contingency tables, so the API is confusing *at first glance*--maybe it will make more sense when I've spent some time with it.

Some time ago, you said

> In reality, I can't imagine a circumstance in which a contingency table would have more than two dimensions but I was just looking ahead for users with strange datasets.

That's just an n-way table, where there are more than two factors.   `chi2_contingency` can handle more than two factors.  There is a four-way example in the `chi2_contingency` docstring.

dtemkin commented:
@WarrenWeckesser Thanks for taking the time to review it. You are right that all of the measures are intended to handle 2d matrices. If it helps, this is the way I handled the issue of an nd-matrix: 

consider an k-dimensional matrix with shape (r, c, k). I recursively loop through the input matrix looking for sub-matrices of size (r, c). Basically a recursive traversal across the number of dimensions. The specified metric is then run on each of these (r, c) matrices , and then the results are reshaped so that it conforms with k. The function 'find_subarrays' achieves this.

If there is an issue with my treatment of n-dimensional, I am more than happy to change this such that the function will only accept 2d matrices but I think my handling is correct, I am not condensing any information, I am just considering that each column is already representative of itself and each dimension.

I hope this helps let me know if there is anything I need to change or alter before it can be merged.

tylerjereddy commented:
Seems like positive feeback from reviewers here, but still some potentially complex discussion, 500 new lines of code, and a long period of inactivity.

dtemkin commented:
Hello all, 

I have made some major changes to my PR. I drop the component which allowed the user to pass a nd-matrix to the function and now I just accept n_obs, n_rows, n_cols, and chi2_stat as function parameters and use those to determine the specified association stat. I also eliminated the bias correction function due to a lack of uniform support and additional complexity without commensurate benefits. Anyway, I hope this makes it easier to approve my PR. 

Let me know if there are any other changes I can make to move the process along.

Thanks

Dan

rgommers commented on scipy/stats/contingency.py:
```diff
@@ -7,9 +7,10 @@
 
 
-__all__ = ['margins', 'expected_freq', 'chi2_contingency']
+__all__ = ["chi2_contingency"]
```
this change looks incorrect; `__all__` should contain all public functions in this file

dtemkin replied:
I made this change before because adding them to __all__ required that I change the stats.__init__.py reference which listed the modules as 
`
chi2_contingency
contingency.margins
contingency.expected_freq
contingency.association
`
Adding the modules to all seemed to require that I drop the contingency on the last three. I made the change but let me know if I should change it back.

rgommers replied:
The reason is that you added `from .contingency import *`, I think you should undo that change.

tylerjereddy replied:
Still a valid request for change here, I think.

rgommers commented on scipy/stats/contingency.py:
```diff
@@ -273,3 +275,124 @@ def chi2_contingency(observed, correction=True, lambda_=None):
+def association(n_obs, n_rows, n_cols, chi2_stat, stat="V"):
+    """Calculates degree of association between variables that are nominal or greater.
+
+    Allows for specification of one of four related methods, Tschuprow's T, Pearson's Contingency Coefficient,
```
style: could you break lines at 80 characters?

rgommers commented on scipy/stats/contingency.py:
```diff
@@ -273,3 +275,124 @@ def chi2_contingency(observed, correction=True, lambda_=None):
+    0.23382
+
+
+    Notes
```
`Notes` should go above `References`

rgommers replied:
Also, please put ` blank line above a section header, and no blank lines below

rgommers commented on scipy/stats/contingency.py:
```diff
@@ -273,3 +275,124 @@ def chi2_contingency(observed, correction=True, lambda_=None):
+    their association. This differs from correlation, although many often mistakenly consider them equivalent.
+    Correlation measures in what way two variables are related, whereas, association measures
+    how related the variables are. As such, association does not subsume independent variables, and is
+    rather a test of independence. Where a value of 1.0 = perfect association or dependent variables, and
```
`=` --> `means`

rgommers commented on scipy/stats/contingency.py:
```diff
@@ -273,3 +275,124 @@ def chi2_contingency(observed, correction=True, lambda_=None):
+    except ValueError:
+        raise ValueError("n_obs, n_cols, n_rows must be int or float")
+
+    if stat.casefold() not in ['t', "c", 'v', 'phi']:
```
can you use `lower` instead? it's clearer and completely identical here. also further down

dtemkin replied:
Made the change as you requested.

rgommers commented on scipy/stats/tests/test_contingency.py:
```diff
@@ -198,3 +198,90 @@ def test_chi2_contingency_bad_args():
 
+
+def test_bad_association_args():
+
```
nit: can you remove blank lines below function definitions in this file?

rgommers commented on scipy/stats/contingency.py:
```diff
@@ -273,3 +275,124 @@ def chi2_contingency(observed, correction=True, lambda_=None):
+    >>> num_cols = obs.shape[1]
+    >>> chi2 = tuple(chi2_contingency(obs))[0]
+
+    Pearson's contingency coefficient
```
this needs a blank line between a line of text and the next `>>>` line to render correctly in the html docs. and same further down

rgommers commented on scipy/stats/contingency.py:
```diff
@@ -273,3 +275,124 @@ def chi2_contingency(observed, correction=True, lambda_=None):
+
+    Parameters
+    ----------
+    stat : {"V", "T", "C", "phi"} (default = "V")
```
This parameter comes last in the signature, hence so should the description.

rgommers replied:
Also, for clarity I'd much prefer more explicit names. For example:
```
`{'pearson', 'cramer', 'tschuprow', 'phi'}`
```
Now it's a single letter without an explanation, hard to read code that way

dtemkin replied:
I made this change but I also allow for the user to specify, (c, t, v) if they wish. It is just a 'hidden' feature (the docs and errors will instruct them to use [cramer, tschuprow, pearson or phi]

rgommers commented on scipy/stats/contingency.py:
```diff
@@ -273,3 +275,124 @@ def chi2_contingency(observed, correction=True, lambda_=None):
     return chi2, p, dof, expected
+
+
+def association(n_obs, n_rows, n_cols, chi2_stat, stat="V"):
```
Wouldn't it be better to have this function simply take `observed` as a single parameter? Looking at the tests, all these parameters are derived from a single set of observations, so why not save the user 4 lines and calculate them inside the function?

@josef-pkt made the same comment in an earlier review.

dtemkin replied:
So, that was one of the changes I made in this latest PR. The main reason I removed it was because there was a lot of contention over how to calculate the stat for an Nd matrix say for one that is 4d like (2x2x5x3). The way I had it was that it would produce an output that was 5x3 matrix of association stats (in this case). 

I am willing to put it back but honestly it seemed like it was much less hassle this way, and it can always be expanded to suit that purpose in a future PR.  

Another solve is to set the function to reject anything that isn't 2d.

Anyway, let me know what you think I should do.

rgommers replied:
Okay, I missed that part of the discussion. I think the concerns of @josef-pkt and @WarrenWeckesser are for >2-D arrays, and when you change to `n_rows, `n_cols` you're anyway assuming 2-D only right? So that would be easier to achieve by keeping `observed` but simply raising an error for >2-D input.

Let's see what the other reviewers say.

rlucas7 commented:
Not familiar with the PR so difficult for me to say. I’ve flagged for follow up this weekend if no one gets to it before then. Thanks for the ping. 

Sincerely, 

-Lucas Roberts 

> On May 19, 2020, at 7:36 PM, Tyler Reddy <notifications@github.com> wrote:
> 
> ﻿
> @tylerjereddy commented on this pull request.
> 
> @WarrenWeckesser @josef-pkt @rlucas7 Does this PR stand a chance of getting merged before we branch in ~1 week?
> 
> In scipy/stats/contingency.py:
> 
> >  
>  
> -__all__ = ['margins', 'expected_freq', 'chi2_contingency']
> +__all__ = ["chi2_contingency"]
> Still a valid request for change here, I think.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

tylerjereddy commented:
This PR has been inactive for 6 months--I'll likely have to bump the milestone if the activity does not pick up soon.

dtemkin reviewed (commented):
Completed changes to __all__ and changed main function back to version with fewer arguments

dtemkin commented:
@tylerjereddy @josef-pkt @rgommers  Sorry this has taken me so long I fixed the code associated with the changes requested. I am still getting an error with azure (due to some formatting errors in the doc string of scipy.optimize.linprog) and circleci (not sure what is going on here).  I am not certain what I can do to address these problems but if there are any further changes that need to be made please dont hesitate to reach out.

rgommers commented:
Looks like this is going to need some git surgery, a number of unrelated changes crept in.

rgommers commented:
@dtemkin it looks like your commits don't need to be preserved, maybe you could just start a fresh branch based on current master, recommit all your changes there, and push it back to this PR?

dtemkin commented:
@rgommers So I totally get the function of what you are asking I am just unclear of how to execute it. Are you saying I should make a new branch with scipy:master and then merge in my changes? Wouldn't all the commits follow me from my current base to the new branch? 

I apologize again for being a rube I am not very good with git.

rgommers commented:
> Are you saying I should make a new branch with scipy:master and then merge in my changes? Wouldn't all the commits follow me from my current base to the new branch?

Indeed. What I was suggesting was just creating a fresh branch, and recommitting all your changes. Either just copy them manually, or (starting on this branch):
```
git reset --mixed master
git add <files with changes you want to keep>
git checkout -b assoc-measures
git commit
git push origin assoc-measures:resolve_conflicts
```
That would get rid of all existing commits, and update this PR to contain just a single new commit with all the code changes you want.

rgommers commented:
continued in gh-13153
